### ISSUE TITLE
Clarification of "last updated (or created)"

### ISSUE THREAD

**Issue by julius-yang** [ocpi/ocpi#428]
tldr: OCPI 2.2.1 section 8.2.1.1 is unclear on what `updated` means.

My interpretation of the `last_updated` field in OCPI classes is that the timestamp should change if and only if some data changed in that field.  For example if EVSE.status changes from AVAILABLE to CHARGING, `EVSE.last_updated` and `Location.last_updated` should become more recent.

However I am pretty sure a partner CPO is setting this field to the current time even if no data has been changed, because they are literally updating it with the same data that existed before.

I think it is ambiguous what "updated" means.  If you apply your current data to the Location/EVSE/Connector object you have technically updated it, even if the data is the same.

If I want to fetch only locations changed since the last hour, `GET /locations?date_from=foobar` is what I send; I expect a subset of the CPO's data, but in this CPO's case, it returns all locations because they have updated all their data in the last hour.

Please have the spec clearly define how to fetch only locations where the data has *changed* since a timestamp.

**Comment by reinierl** [comment]
I agree with you @julius-yang that this is not how the `last_updated` field is intended.

On the other hand there is no hard wording that bans what they are doing outright.

In OCPI 3.0 data replication will work differently anyway and I hope the general architecture of it will coach people away from coupling OCPI datasets to tightly to internal representations of the EVSEs and other objects.

Do you expect immediate action from us or is it enough to know that this will be changed in OCPI 3.0?

**Comment by julius-yang** [comment]
I need to understand what the intent of `last_updated` and GET /locations?start_from=foo is.  This allows me to ask a partner CPO whether their implementation is:

1. implemented out of spec
2. implemented within spec but out of spec's intent
3. implemented within spec and with the spec's intent ... incorrectly (not quite the same as a bug, if you see what I mean)

If I understand you correctly, the intent of the spec is that `last_updated` should only change if any values of the Location/EVSE/Connector change.  In addition if a Connector changes any value, then last_updated should be changed on Location/EVSE/Connector; if an EVSE changes a value, then last_updated should be different on Location/EVSE.

If that is correct I can take that to the CPO and discuss, and you can close this issue.  Thanks!

**Comment by reinierl** [comment]
> If I understand you correctly, the intent of the spec is that last_updated should only change if any values of the Location/EVSE/Connector change. In addition if a Connector changes any value, then last_updated should be changed on Location/EVSE/Connector; if an EVSE changes a value, then last_updated should be different on Location/EVSE.

That is correct indeed. In the Locations module it says this:

> If the optional parameters {date_from} and/or {date_to} are provided, only Locations with (last_updated) between the given {date_from} (including) and {date_to} (excluding) will be returned. In order for this to work properly, the following logic MUST be implemented accordingly: If an EVSE is updated, also the 'parent' Location’s last_updated field needs to be updated. Similarly, if a Connector is updated, the EVSE’s last_updated and the Location’s last_updated fields need to be updated.

**Comment by julius-yang** [comment]
> > If I understand you correctly, the intent of the spec is that last_updated should only change if any values of the Location/EVSE/Connector change. In addition if a Connector changes any value, then last_updated should be changed on Location/EVSE/Connector; if an EVSE changes a value, then last_updated should be different on Location/EVSE.
> 
> That is correct indeed. In the Locations module it says this:
> 
> > If the optional parameters {date_from} and/or {date_to} are provided, only Locations with (last_updated) between the given {date_from} (including) and {date_to} (excluding) will be returned. In order for this to work properly, the following logic MUST be implemented accordingly: If an EVSE is updated, also the 'parent' Location’s last_updated field needs to be updated. Similarly, if a Connector is updated, the EVSE’s last_updated and the Location’s last_updated fields need to be updated.

Thanks.  In that case "If an EVSE is updated" actually means "If an EVSE's values have changed" rather than "If an EVSE's values have been written to".

I will say however that those two events are distinct and both are useful.  Here's two use cases:

1) A cron job iterates over EVSEs and pulls status from them.  It writes the EVSE status to the db, even if the status is the same.  "last_written" is changed to now(), but "last_changed" does not change.
2) A cron job iterates over EVSEs and pulls status from them.  It writes the EVSE status to the db where the status is different.  "last_written" is changed to now(), and "last_changed" is changed to now().

This distinction allows CPOs/eMSPs to tell whether the EVSE is being written to (the cron job is operating correctly), and when the EVSE last changed state (perhaps it has only reported CHARGING for the last 7 days, rather than the cron job being broken).  And of course it enables eMSPs to pull a delta -- only Locations/EVSEs that have changed.

**Comment by ahzf** [comment]
@julius-yang Most implementations I'm aware of ignore this very awful design flaw of OCPI and handle EVSE status changes in a seperate way just like other EV roaming protocols do, e.g. Hubjects OICP.

In computer science a change tracking timestamp must not change when a property "changes" from one value to the exact same value. The entire mechanism is for _change tracking_, not for detecting if the system is still alive, not for "refreshing" something, or anything else non-CS guys come up with.

Why connectors have their own change tracking is another **wtf** in OCPI. The main object should be called "charging station", have 1 or multiple EVSE(s) and those again have 1 or multiple connector(s). Only at the charging station level there should be a change tracking.

Remember also that OCPI uses this "merge PATCH" strategy thus only top-level properties can be changed easily. Embedded data structures can only be changed as bulk. The CS-correct way to solve this is to use JSON PATCH documents with "add property, change property, and delete property" commands. This OCPI work-around with nested change tracking is a bug pattern.

Keeping this in mind the best way to implement the EVSE status is to create a special HTTP handler and keep the status history in dedicated data structures. So do not keep the EVSE status as a property of a JSON document.

All in all because of this design flaws OCPI is really a pain to implement and test... it's time to deprecate it.

**Comment by mbayings** [comment]
@ahzf This Github is meant to come with useful comments for OCPI. We will not accept swear words. If you can't contribute in a positive way you will be blocked. There is already enough blaming and shaming and all kind of nasty messages on other social media. Please give useful normal input or don't do it at all. We accept comments and criticises, but not using swear words and other words that are not helpful and supporting the protocol. Critics should lead to improvements. What you do is only showing how negative you are. This is not the place for that.
